test/infrastructure/docker/exp/internal/controllers/dockermachinepool_controller.go

```diff
@@ -230,25 +299,232 @@ func (r *DockerMachinePoolReconciler) reconcileNormal(ctx context.Context, clust
+		dockerMachinePool.Status.Ready = true
+		conditions.MarkTrue(dockerMachinePool, expv1.ReplicasReadyCondition)
+
+		return ctrl.Result{}, nil
+	}
+
+	dockerMachinePool.Status.Ready = false
+	conditions.MarkFalse(dockerMachinePool, expv1.ReplicasReadyCondition, expv1.WaitingForReplicasReadyReason, clusterv1.ConditionSeverityInfo, "")
 
 	// if some machine is still provisioning, force reconcile in few seconds to check again infrastructure.
-	if !dockerMachinePool.Status.Ready && res.IsZero() {
+	if res.IsZero() {
 		return ctrl.Result{RequeueAfter: 10 * time.Second}, nil
 	}
+
 	return res, nil
 }
 
+func getDockerMachines(ctx context.Context, c client.Client, cluster clusterv1.Cluster, machinePool expv1.MachinePool, dockerMachinePool infraexpv1.DockerMachinePool) (*infrav1.DockerMachineList, error) {
+	dockerMachineList := &infrav1.DockerMachineList{}
+	labels := map[string]string{
+		clusterv1.ClusterNameLabel:     cluster.Name,
+		clusterv1.MachinePoolNameLabel: machinePool.Name,
+	}
+	if err := c.List(ctx, dockerMachineList, client.InNamespace(dockerMachinePool.Namespace), client.MatchingLabels(labels)); err != nil {
+		return nil, err
+	}
+
+	return dockerMachineList, nil
+}
+
+// DeleteOrphanedDockerMachines deletes any DockerMachines owned by the DockerMachinePool that reference an invalid providerID, i.e. not in the latest copy of the node pool instances.
+func (r *DockerMachinePoolReconciler) DeleteOrphanedDockerMachines(ctx context.Context, cluster *clusterv1.Cluster, machinePool *expv1.MachinePool, dockerMachinePool *infraexpv1.DockerMachinePool, nodePoolMachineStatuses []docker.NodePoolMachineStatus) error {
+	log := ctrl.LoggerFrom(ctx)
+	log.V(2).Info("Deleting orphaned machines", "dockerMachinePool", dockerMachinePool.Name, "namespace", dockerMachinePool.Namespace, "nodePoolMachineStatuses", nodePoolMachineStatuses)
+	dockerMachineList, err := getDockerMachines(ctx, r.Client, *cluster, *machinePool, *dockerMachinePool)
+	if err != nil {
+		return err
+	}
+
+	log.V(2).Info("DockerMachineList kind is", "kind", dockerMachineList.GetObjectKind())
+
+	nodePoolMachineNames := map[string]struct{}{}
+	for _, nodePoolMachineStatus := range nodePoolMachineStatuses {
+		nodePoolMachineNames[nodePoolMachineStatus.Name] = struct{}{}
+	}
+
+	for i := range dockerMachineList.Items {
+		dockerMachine := &dockerMachineList.Items[i]
+		if _, ok := nodePoolMachineNames[dockerMachine.Name]; !ok {
+			machine, err := util.GetOwnerMachine(ctx, r.Client, dockerMachine.ObjectMeta)
+			// Not found doesn't return an error, so we need to check for nil.
+			if err != nil {
+				return errors.Wrapf(err, "error getting owner Machine for DockerMachine %s/%s", dockerMachine.Namespace, dockerMachine.Name)
+			}
+			if machine == nil {
+				// If MachinePool is deleted, the DockerMachine and owner Machine doesn't already exist, then it will never come online.
+				if mpDeleted := isMachinePoolDeleted(ctx, r.Client, machinePool); mpDeleted {
```

Comment:
is this code required? If the machinepool is deleted, since we have set owner reference of infra machinepool in the machinepool machine, it will anyways get automatically deleted right? Thats what my tests show, can you please confirm if that understanding is not correct.

Reply:
That's a good point, I believe you would be correct here as long as the ownerRef is set. In that case, I think we can probably drop this check. I'll tag @fabriziopandini to verify as it was his original suggestion.

Reply:
I think it should no longer be needed and made some changes in refactoring.